### DISCUSSION TITLE
How to analyst my 2 options in the nextjs 13 app router?

### DISCUSSION THREAD

**Discussion by tofuwu-dev** [vercel/next.js#55529] [Help] [answered]
### Summary

Hi @everyone,

I use nextjs 13 app router.
I would like to discuss the following with you:

1. Page is `Server Component` as a default. But I am importing 5+ client components in the page.
2. Page is `Server Component` as a default but I declared it to `Client Component` using `use client` directive and I am importing 5+ server components in the page.

What is better to performance? How to analyst exactly?

In my experiment, 1st option was better than 2nd option for performance but code was more neat for 2nd option.

Expect your responses!

**Comment by icyJoseph** [comment, marked as answer]
I think number 2 is kind of problematic for various reasons.

First, it opts in the entire page into having a JS bundle, which sort of, defeats one of the purposes of server components, which is to reduce JS bundle size. As a rule of thumb, client components should be moved to the leaves of the React component tree.

Second, importing components into a `use client` tree, makes an assumption. That none of these is using `async/await`. Since a component is by default a server component, unless `use client` is used, you are free to use `async/await` and then if you import that directly into a client component, you'll end up with an error, because `Client Components` do not support Promise-JSX.

Third, there's a section in the docs, regarding, supported and unsupported patterns to piece together client and server components:

- https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#interleaving-server-and-client-components

There you can see this snippet:

```tsx
'use client'
 
// You cannot import a Server Component into a Client Component.
import ServerComponent from './Server-Component'

```

So please avoid that. 

In terms of performance, from what you describe, is a reduced JS bundle. Just measure how much JS data is transferred to your browser, by looking at the network tab.

You can also use this tool, (adapting it your needs): https://github.com/jherr/njs13-n-plus-one/blob/main/tester/index.js

It'll print something like:

```
            HTML: 5.97 KB - 38.0%
            JSON: 9.75 KB - 62.0%
JSON Client Data: 0.00 KB - 0.0%
```

Here you can see how the author uses it: https://youtu.be/FfHsIio4aCU?t=1135

**Reply by tofuwu-dev** [reply]
@icyJoseph Thanks for your suggestion!

But I think there are some difficulties to use Server Components freely because we mostly use hook like `useState` and `useEffect`.

Anyways, I learnd so much from you this time and I know you are a good person as well as senior developer.

Thanks again!

**Reply by tofuwu-dev** [reply]
Is there any ways to get query router parasm in server components btw? @icyJoseph 
Because I need to declare `use client` directive to do router manipulation - like `useRouter`, `usePathname` and `useSearchParams`

**Reply by icyJoseph** [reply]
That's one thing that people have been missing. Currently you have to prop drill from the page server component down to the child that needs it. AFAIK, there's other tricks but nothing that's widely accepted yet.

**Reply by icyJoseph** [reply]
Composition too, I forgot to mention. 

And I once saw an npm package, I'll try to find it and post back here.

**Reply by icyJoseph** [reply]
Here's a link with some info: https://github.com/vercel/next.js/discussions/50405#discussioncomment-6670717

- https://www.npmjs.com/package/server-only-context

⚠️ I have not had the chance to try it out myself ⚠️⚠️⚠️
